pkg/account/component.go

```diff
@@ -223,32 +236,43 @@ func (c *component) UpdateAccount(ctx context.Context, user api.UpdatableAccount
-		user.Nationality, oldUser.Nationality,
-		user.IDDocumentType, oldUser.IDDocumentType,
-		user.IDDocumentNumber, oldUser.IDDocumentNumber,
-		user.IDDocumentExpiration, oldUser.IDDocumentExpiration,
-		user.IDDocumentCountry, oldUser.IDDocumentCountry,
-	)
-
-	// profileUpdated: Add here all fields which are not accreditation-dependant...
-	// phone number is not added here as there is already an email sent for phone number verification
-	// (currently no other field)
-	var profileUpdated = revokeAccreditations
+	var fieldsComparator = fields.NewFieldsComparator().
+		CompareValueAndFunctionForUpdate(fields.Email, user.Email, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.FirstName, user.FirstName, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.LastName, user.LastName, oldUserKc.GetFieldValues).
+		CompareOptionalAndFunction(fields.BusinessID, user.BusinessID, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.Gender, user.Gender, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.PhoneNumber, user.PhoneNumber, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.BirthDate, user.BirthDate, oldUserKc.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.BirthLocation, user.BirthLocation, oldUser.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.Nationality, user.Nationality, oldUser.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.IDDocumentType, user.IDDocumentType, oldUser.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.IDDocumentNumber, user.IDDocumentNumber, oldUser.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.IDDocumentExpiration, user.IDDocumentExpiration, oldUser.GetFieldValues).
+		CompareValueAndFunctionForUpdate(fields.IDDocumentCountry, user.IDDocumentCountry, oldUser.GetFieldValues)
+
+	var newAccreditations []string
+	newAccreditations, err = c.accreditationEvaluator.EvaluateAccreditations(ctx, realm, userID, fieldsComparator, oldUserKc.GetFieldValues(fields.Accreditations))
+	if err != nil {
+		// Message already logged
+		return err
+	}
+	oldUserKc.SetFieldValues(fields.Accreditations, newAccreditations)
+
+	// profileChangesCount
+	var profileChangesCount = len(fieldsComparator.UpdatedFields())
 
 	// manage email change
 	var prevEmail *string
-	if isUpdated(user.Email, oldUserKc.Email) {
+	if fieldsComparator.IsAnyFieldUpdated(fields.Email) {
 		prevEmail = oldUserKc.Email
 		oldUserKc.SetAttributeString(constants.AttrbEmailToValidate, *user.Email)
 		actions = append(actions, ActionVerifyEmail)
+		profileChangesCount-- //
 	}
 
 	// manage phone number change
-	if isUpdated(user.PhoneNumber, oldUserKc.GetAttributeString(constants.AttrbPhoneNumber)) {
+	if fieldsComparator.IsAnyFieldUpdated(fields.PhoneNumber) {
 		oldUserKc.SetAttributeString(constants.AttrbPhoneNumberToValidate, *user.PhoneNumber)
 		actions = append(actions, ActionVerifyPhoneNumber)
```

Comment:
profileChangesCount-- missing ?

Reply:
No, this variable is only used to check if we have to send an email for a change of the profile. But we do not want to take into account the email. But in case of a phone number change, we want to send the profile email change

Reply:
Sorry, I do not understand it ... and I have the feeling that it is becoming too much complicated to maintain it without risk of introduction of regressions.

Reply:
There are specific email sent for email changes... So for the "profile changes email", we dont take email changes into account ; that's why we decrement profileChangesCount.
For phone number, there is no specific email sent... so phone number do need to be taken into account in profileChangesCount